File: November2020/JumpGameIII.java
class JumpGameIII {
    public boolean canReach(int[] arr, int start) {
        // O(n) : TC
        // O(n) : SC boolean array
        Queue<Integer> qu = new LinkedList<>();
        boolean[] visited = new boolean[arr.length];
        qu.offer(start);
        while(!qu.isEmpty()){
            int size = qu.size();
            while(size-->0){
                Integer head = qu.poll();
                if(arr[head] == 0){
                    return true;
                }
                if(!visited[head]){
                    int forwardIndex = head + arr[head];
                    if(forwardIndex>=0 && forwardIndex <arr.length){
                        qu.offer(forwardIndex);

                    }
                    int backwardIndex = head - arr[head];
                    if(backwardIndex>=0 && backwardIndex <arr.length){
                        qu.offer(backwardIndex);

                    }
                    visited[head] = true;
                }
            }
        }

        return false;

    }
}

File: December2021/C++/Jump_Game_III.cpp
class Solution {
    
public:
    bool canReach(vector<int>& arr, int start) {
        vector<pair<bool,bool>> visited(arr.size(),{false,false});
        return  dfs(arr,start,visited);
    }
    
    bool dfs(vector<int>& arr,int idx,vector<pair<bool,bool>>& visited)
    {
        if(idx<0 || idx>=arr.size()) return false;
        
        if(arr[idx]==0) return true;
        
        if(visited[idx].first)
            return visited[idx].second;
        
        visited[idx].first=true;
        visited[idx].second= dfs(arr,idx+arr[idx],visited) || dfs(arr,idx-arr[idx],visited);
        return visited[idx].second;
    }
};